low-level/files/opt/seagate/sspl/setup/health_provider/storage.py

```diff
@@ -0,0 +1,153 @@
+# (at your option) any later version.
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com
+
+"""
+ ***************************************************************************
+  Description: Storage class to provide fru health and specific information
+               through controller web service.
+ ***************************************************************************
+"""
+
+import errno
+import json
+import re
+import time
+
+from framework.utils.conf_utils import (
+    GLOBAL_CONF, SRVNODE, Conf, STORAGE_TYPE_KEY, NODE_TYPE_KEY, SITE_ID_KEY,
+    RACK_ID_KEY, NODE_ID_KEY, CLUSTER_ID_KEY, RELEASE, TARGET_BUILD)
+from error import HealthProviderError
+from resource_health import ResourceHealth
+
+
+class StorageHealth(ResourceHealth):
+    """Provides health information of FRUs in storage"""
+
+    name = "storage"
+
+    def __init__(self):
+        """Initialize storage"""
+        super().__init__()
+        self.validate_storage_type_support()
+        self.storage_frus = {
+            "controllers": self.get_controllers_info
+            }
+
+    @staticmethod
+    def validate_storage_type_support():
+        """Check for supported storage type"""
+        storage_type = Conf.get(GLOBAL_CONF, STORAGE_TYPE_KEY, "virtual").lower()
+        supportted_types = ["5u84", "rbod", "pods"]
+        if storage_type not in supportted_types:
+            raise HealthProviderError(
+                errno.EINVAL,
+                "Health provider is not supported for storage type '%s'." % storage_type)
+
+    @staticmethod
+    def get_node_details(node):
+        """
+        Parse node information and returns left string and instance.
+        Example
+            "storage"    -> ("storage", "*")
+            "storage[0]" -> ("storage", "0")
+        """
+        res = re.search(r"(\w+)\[([\d]+)\]|(\w+)", node)
+        inst = res.groups()[1] if res.groups()[1] else "*"
+        node = res.groups()[0] if res.groups()[1] else res.groups()[2]
+        return node, inst
+
+    def get_health_info(self, rpath):
+        """
+        Fetch health information for given FRU
+        rpath: Resouce path (Example: hw>controllers)
+        """
+        info = {}
+        nodes = rpath.strip().split(">")
+        leaf_node, _ = self.get_node_details(nodes[-1])
+        if leaf_node == "storage":
+            info["last_updated"] = int(time.time())
+            for fru in self.storage_frus:
+                info.update({fru: self.storage_frus[fru]()})
```

Comment:
**Great, this good approach, as we had done elsewhere in SSPL also. Keeps it modular and allow people to contribute code without much disturbance**